TypeScript error: "tagName" does not exist in type 'ClassDeclaration'

I originally opened this issue on the custom-elements-manifest analyzer (https://github.com/open-wc/custom-elements-manifest/issues/43), but I'm told this is a bug in the spec, so reopening here.

In the [custom-elements-manifest analyzer playground](https://custom-elements-manifest.netlify.app/), use this custom element:

```js
class MyElement extends HTMLElement {}
customElements.define("my-element", MyElement)
```

You'll get this JSON:

<details><summary>click to expand</summary>

```json
{
  "schemaVersion": "1.0.0",
  "readme": "",
  "modules": [
    {
      "kind": "javascript-module",
      "path": "src/my-element.js",
      "declarations": [
        {
          "kind": "class",
          "description": "",
          "name": "MyElement",
          "superclass": {
            "name": "HTMLElement"
          },
          "tagName": "my-element",
          "customElement": true
        }
      ],
      "exports": [
        {
          "kind": "custom-element-definition",
          "name": "my-element",
          "declaration": {
            "name": "MyElement",
            "module": "src/my-element.js"
          }
        }
      ]
    }
  ]
}
```

</details>

If you compare this to the [schema](https://github.com/webcomponents/custom-elements-manifest/blob/2e391f2993431c231c4d2f34ce6d2f1d24ddad69/schema.d.ts), it doesn't match because of `tagName` on the `ClassDeclaration`.

You can repro using a simple TypeScript file:

<details><summary>click to expand</summary>

```ts
import { Package } from 'custom-elements-manifest/schema'

const json: Package = {
  "schemaVersion": "1.0.0",
  "readme": "",
  "modules": [
    {
      "kind": "javascript-module",
      "path": "index.js",
      "declarations": [
        {
          "kind": "class",
          "description": "",
          "name": "MyElement",
          "superclass": {
            "name": "HTMLElement"
          },
          "tagName": "my-element",
          "customElement": true
        }
      ],
      "exports": [
        {
          "kind": "custom-element-definition",
          "name": "my-element",
          "declaration": {
            "name": "MyElement",
            "module": "index.js"
          }
        }
      ]
    }
  ]
}
```

</details>

Then run:

```shell
mkdir test
cd test
npm init --yes
npm i --save custom-elements-manifest@1.0.0 typescript@4.3.4
npx tsc index.ts
```

You'll see the error:

```
test.ts:18:11 - error TS2322: Type '{ kind: "class"; description: string; name: string; superclass: { name: string; }; tagName: string; customElement: true; }' is not assignable to type 'Declaration'.
  Object literal may only specify known properties, and '"tagName"' does not exist in type 'ClassDeclaration'.

18           "tagName": "my-element",
             ~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.
```

**Expected behavior**

I would expect the output schema to match the one from `schema.d.ts`.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,15 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 <!-- ### Removed -->
 <!-- ### Fixed -->
 
+<!-- ## [x.y.z] - YYYY-MM-DD -->
+## Unreleased
+<!-- ### Changed -->
+<!-- ### Added -->
+<!-- ### Removed -->
+### Fixed
+
+- Fixed how custom element declarations extend class declarations. Previously CustomElementDeclaration didn't include CustomElement properties, and MixinDeclaration required some CustomElement properties. Added new CustomElementMixinDeclaration interface. Fixes https://github.com/webcomponents/custom-elements-manifest/issues/69
+
 ## [1.0.0] - 2021-06-10
 
 Initial release

--- a/schema.d.ts
+++ b/schema.d.ts
@@ -131,18 +131,19 @@ export type Declaration =
   | FunctionDeclaration
   | MixinDeclaration
   | VariableDeclaration
-  | CustomElementDeclaration;
+  | CustomElementDeclaration
+  | CustomElementMixinDeclaration;
 
 /**
  * A reference to an export of a module.
  *
  * All references are required to be publically accessible, so the canonical
  * representation of a reference is the export it's available from.
- * 
+ *
  * `package` should generally refer to an npm package name. If `package` is
  * undefined then the reference is local to this package. If `module` is
  * undefined the reference is local to the containing module.
- * 
+ *
  * References to global symbols like `Array`, `HTMLElement`, or `Event` should
  * use a `package` name of `"global:"`.
  */
@@ -173,16 +174,18 @@ export interface SourceReference {
  * neccessarily part of a custom element class, but belong to the definition
  * (often called the "registration") or the `customElements.define()` call.
  *
- * Because classes and tag anmes can only be registered once, there's a
+ * Because classes and tag names can only be registered once, there's a
  * one-to-one relationship between classes and tag names. For ease of use,
  * we allow the tag name here.
  *
  * Some packages define and register custom elements in separate modules. In
  * these cases one `Module` should contain the `CustomElement` without a
  * tagName, and another `Module` should contain the
- * `CustomElement`.
+ * `CustomElementExport`.
  */
-export interface CustomElementDeclaration extends ClassDeclaration {}
+export interface CustomElementDeclaration
+  extends ClassDeclaration,
+    CustomElement {}
 
 /**
  * The additional fields that a custom element adds to classes and mixins.
@@ -351,7 +354,7 @@ export interface Type {
    */
   references?: TypeReference[];
 
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 /**
@@ -432,7 +435,7 @@ export interface ClassLike {
   mixins?: Array<Reference>;
   members?: Array<ClassMember>;
 
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 export interface ClassDeclaration extends ClassLike {
@@ -468,7 +471,7 @@ export interface ClassField extends PropertyLike {
   static?: boolean;
   privacy?: Privacy;
   inheritedFrom?: Reference;
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 export interface ClassMethod extends FunctionLike {
@@ -476,7 +479,7 @@ export interface ClassMethod extends FunctionLike {
   static?: boolean;
   privacy?: Privacy;
   inheritedFrom?: Reference;
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 /**
@@ -531,18 +534,24 @@ export interface ClassMethod extends FunctionLike {
  * }
  * ```
  */
-export interface MixinDeclaration extends CustomElement, FunctionLike {
+export interface MixinDeclaration extends ClassLike, FunctionLike {
   kind: 'mixin';
+}
+
+/**
+ * A class mixin that also adds custom element related properties.
+ */
+export interface CustomElementMixinDeclaration extends MixinDeclaration, CustomElement {
 }
 
 export interface VariableDeclaration extends PropertyLike {
   kind: 'variable';
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 export interface FunctionDeclaration extends FunctionLike {
   kind: 'function';
-  source? : SourceReference;
+  source?: SourceReference;
 }
 
 export interface Parameter extends PropertyLike {
@@ -587,5 +596,5 @@ export interface Demo {
    */
   url: string;
 
-  source? : SourceReference;
+  source?: SourceReference;
 }

